loader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/loader/BadRow.scala

```diff
@@ -12,52 +12,142 @@
+
+  /**
+    * Represents the failure case where data can not be parsed as a proper event
+    * @param payload data blob tried to be parsed
+    * @param errors  errors in the end of the parsing process
+    */
+  final case class ParsingError(payload: String, errors: NonEmptyList[String]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.LoaderParsingError, (this: BadRow).asJson)
+  }
+
+  /**
+    * Represents the errors due to something went wrong internally
+    * In this kind of errors, event is parsed properly, therefore
+    * enriched event is given in the error as properly parsed
+    * @param enrichedEvent event which is enriched successfully
+    * @param errors        info of errors
+    */
+  final case class InternalError(enrichedEvent: Event, errors: NonEmptyList[InternalErrorInfo]) extends BadRow {
+    def getSelfDescribingData: SelfDescribingData[Json] =
+      SelfDescribingData(BadRowSchemas.LoaderInternalError, (this: BadRow).asJson)
+  }
 
-  implicit val decoder: Decoder[BadRow] =
-    Decoder.instance[BadRow] { cursor =>
-      for {
-        jsonObject <- cursor.value
-          .asObject
-          .toRight(DecodingFailure("Bad row should be an object", cursor.history))
-        map = jsonObject.toMap
-        line <- map.get(Line)
-          .flatMap(_.asString)
-          .toRight(DecodingFailure(s"$Line should be a string", cursor.history))
-        decodedLine <- Either.catchNonFatal(new String(Base64.getDecoder.decode(line)))
-          .leftMap(error => DecodingFailure(s"Cannot decode $Line ${error.getMessage}", cursor.history))
-        errors <- map.get(Errors)
-          .flatMap(_.as[NonEmptyList[String]].toOption)
-          .toRight(DecodingFailure(s"Cannot decode $Errors", cursor.history))
-
-      } yield BadRow(decodedLine, errors)
+  implicit val badRowCirceJsonEncoder: Encoder[BadRow] =
+    Encoder.instance {
+      case ParsingError(payload, errors) => json"""{
+          "payload": $payload,
+          "errors": $errors,
+          "processor": $BQLoaderProcessorInfo
+        }"""
+      case InternalError(enrichedEvent, errors) => json"""{
+          "enrichedEvent": $enrichedEvent,
+          "failures": $errors,
+          "processor": $BQLoaderProcessorInfo
+        }"""
     }
+
+  /**
+    * Gives info about the reasons of the internal errors
+    */
+  sealed trait InternalErrorInfo
+
+  object InternalErrorInfo {
+
+    /**
+      * Represents errors which occurs when making validation
+      * against a schema with Iglu client
+      * @param schemaKey instance of schemaKey which could not be found
+      * @param error     instance of ClientError which gives info
+      *                  about the reason of lookup error
+      */
+    final case class IgluValidationError(schemaKey: SchemaKey, error: ClientError) extends InternalErrorInfo
```

Comment:
I don't think we can say that Iglu validation error is "internal" error. Quite othwerwise, schemas and events are best examples of external errors.

Reply:
I think I made some mistake in there related with the 'internal' keyword. I will change all of them. What do you think would be best ? Do you think splitting them as `LoaderParsingError`, `LoaderError`, `RepeaterParsingError` and `RepeaterError` is okay ? In other words, I will remove all the occurrences of the `internal` ?

Reply:
Hypothetically, `LoaderParsingError` is subset of `LoaderError`. Otherwise sounds a lot better.

Reply:
Yeah you are right, I've used `internal` keyword in order to differentiate them but as you suggest, its usage is not exactly right. Do you have anything in your mind which we can use for errors different than parsing error ?

Reply:
The more I think about it the more I relise there's something wrong with this structure.

1. `errors: NonEmptyList[InternalErrorInfo]` tells us that we can have heterogenius list of errors, i.e. iglu errors and cast errors. Is it really possible? They seem to be different steps in the transformation? It can be either `NEL[ClientError]` OR NEL[CastError]`. Correct me if I'm wrong.
2. If they're really separate, is there any reasons to combine them into a single type?
3. If we split them (which is just a technically correct solution) we won't have a problem with naming:

```
trait BadRow
class ParsingError(payload: String, errors: NonEmptyList[String]) extends BadRow
class IgluError(enrichedEvent: Event, errors: NonEmptyList[(SchemaKey, ClientError)]) extends BadRow // tuple is just for example
class CastError(enrichedEvent: Event, errors: NonEmptyList[(Json, SchemaKey, NonEmptyList[BigQueryCastError])]) extends BadRow
```

Am I missing something?

Reply:
Yeah, I think it makes sense. However, in that case, we are differing with the structure of Snowflake Loader bad rows. Not saying this is bad thing, I just want to point it out. So, we will create two different schema instead of four which is one for loader bad rows and one for repeater bad row ?

Reply:
Also, we will repeat `enrichedEvent` for every new type of non parsing error. I think it was the reason why we tried to make one type error which is the non parsing error ones.

Reply:
> we are differing with the structure of Snowflake Loader

I don't see how it will be different. Even if it will - semantics are also completely different (Snowflake doesn't have Iglu neither casting).

> two different schema instead of four 

Why not six? E.g. `loader_parsing_error` (existing) + `iglu_error` (common with other loaders) + `bigquery_cast_error` (the only one so far specific for BQ Loader) + some for repeater

Reply:
> Also, we will repeat enrichedEvent for every new type of non parsing error

I don't see it as an issue (unless I miss something).